Fix elastic-agent cloud image run issues

fearful-symmetry opened the pull request:
## What does this PR do?

We're seeing issues with various docker builds of elastic-agent failing with an `"input not supported"` error. It appears that when elastic-agent starts up in container mode, it'll set the paths incorrectly under V2, so we aren't reading the specfiles from the component directory. 

This makes a few changes to the container setup to get the paths working under the "flat" V2 structure. I'm not 100% sure what else is happening under "container mode" so I want someone with a little more experience here to make sure I'm not breaking anything. I tested this natively and inside a container, and it seems to work.

In the future, we might want to modify the specfile `load.go` so we return an specific error if we can't find any specfiles

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

elasticmachine commented:
## :grey_exclamation: Build Aborted
  
  > There is a new build on-going so the previous on-going builds have been aborted.
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1630/detail/PR-1630/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1630/detail/PR-1630/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1630/detail/PR-1630/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/elastic-agent%2Felastic-agent-mbp%2FPR-1630/detail/PR-1630/1//artifacts)   [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2022-10-27T22:24:25.478Z&rangeTo=2022-10-27T22:44:25.478Z&transactionName=BUILD+elastic-agent%2Felastic-agent-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=28aa89500d3730a67859e69eee630411&transactionId=57832c10366278ed)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-10-27T22:34:25.478+0000


* Duration: 3 min 21 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->
















## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages.


- `run integration tests` : Run the Elastic Agent Integration tests.


- `run end-to-end tests` : Generate the packages and run the E2E Tests.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 98.305% (`58/59`) | :grey_exclamation: 
Files | 70.256% (`137/195`) | :grey_exclamation: 
Classes | 70.161% (`261/372`) | :grey_exclamation: 
Methods | 54.318% (`780/1436`) | :grey_exclamation: 
Lines | 39.579% (`8276/20910`) | :grey_exclamation: 
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

cmacknz commented:
The biggest thing we need to make sure we don't break is using custom a STATE_PATH, which is often used to mount that directory on a persistent or host volume when the agent runs on Kubernetes so that the directory contents persist across restarts and upgrades. It should hopefully be easy to test this by just changing the state path environment variable.

cmacknz commented on internal/pkg/agent/cmd/container.go:
```diff
@@ -789,10 +789,9 @@ func setPaths(statePath, configPath, logsPath string, writePaths bool) error {
 	originalInstall := paths.Install()
 	originalTop := paths.Top()
-	paths.SetTop(topPath)
+	paths.SetTop(statePath) // change this?
```
I think this is correct from my reading of the comments here: https://github.com/elastic/elastic-agent/blob/fc3eba3501a0695a477a46331293331873bb850c/internal/pkg/agent/application/paths/common.go#L59-L71

I also suspect you want to change the `topPath` variable here so that the permissions are set correctly:

https://github.com/elastic/elastic-agent/blob/fc3eba3501a0695a477a46331293331873bb850c/internal/pkg/agent/cmd/container.go#L769-L775

cmacknz commented on internal/pkg/agent/cmd/container.go:
```diff
@@ -789,10 +789,9 @@ func setPaths(statePath, configPath, logsPath string, writePaths bool) error {
+	paths.SetTop(statePath) // change this?
 	paths.SetConfig(configPath)
-	// when custom top path is provided the home directory is not versioned
-	paths.SetVersionHome(false)
```
I'm curious to know what this does (I expect we need Michal or Blake to answer this). It's a bit hard to tell from the code.

As far as I can tell in the container I built, the home path is versioned by default. I get `/usr/share/elastic-agent/data/elastic-agent-fc3eba`. I don't know that this would change if the STATE_PATH were to change either so removing it seems right but it would be good to confirm.

fearful-symmetry replied:
Yah, this just adds the `elastic-agent-fc3eba` directory, which is still there in docker.

fearful-symmetry commented:
@cmacknz , yah, I suspect there's deeper problems with how the state_path is being set, as elastic-agent in its current state (and in this PR) uses the state path for the entire run directory. I assume this code made more sense before the "flattened" V2 directory structure, but right now I'm not sure we have the mechanism to separate out the state info from the rest of the pre-installed elastic-agent assets.

fearful-symmetry commented:
Alright, added the changelog and fixed up a few things. @cmacknz right now, it seems like elastic-agent just kind of mixes statefiles in the flat directory along with everything else. If the intent is to have `STATE_PATH` be a separate directory that users can mount just for tracking state, we may want to open an issue to track that.

michalpristas commented on internal/pkg/agent/cmd/container.go:
```diff
@@ -40,9 +40,9 @@ import (
 	maxRequestRetriesEnv     = "KIBANA_REQUEST_RETRY_COUNT"
-	defaultRequestRetrySleep = "1s"                             // sleep 1 sec between retries for HTTP requests
-	defaultMaxRequestRetries = "30"                             // maximum number of retries for HTTP requests
-	defaultStateDirectory    = "/usr/share/elastic-agent/state" // directory that will hold the state data
```
this is here probably to share state across runs. have mounted volume. 
this PR fixes the problem but i don't believe it's proper fix. 

i believe proper fix would be to fix not just `data/downloads` but also `data/components` 
will prepare PR on monday and test it as well

cmacknz commented:
@michalpristas should this be closed in favour of https://github.com/elastic/elastic-agent/pull/1653?

michalpristas commented:
closing in favor of #1653
